### DISCUSSION TITLE
ResourceDictionary, event handlers and x:Class

### DISCUSSION THREAD

**Discussion by sarosigy** [unoplatform/uno#10094] [Q&A]
Now I'm facing a problem:
Created a ResourceDictionary which is referenced by App.xaml like

```
<ResourceDictionary.MergedDictionaries>
    <local:Bindables />
```
and created an xaml and cs file for it:
```
<ResourceDictionary
    x:Class="JustAStore.Bindables"
```
and
```
namespace JustAStore
{
    public partial class Bindables: ResourceDictionary
```

However, the events on the included DataTemplates are not referenced.
The InitializeComponent is there, I can go to the definition.
When doing this, In the generated code I find warnings:
`// WARNING Unable to use event Click without a backing class (use x:Class)
`
But I have x:Class in the xaml...

So, after some research I found this in the XamlFileGenerator (src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs):

```
			if (!isInsideDataTemplate.isInside)
			{
				writeEvent("");
			}
			else if (_xClassName?.ClassName != null)
			{
				writeEvent(CurrentResourceOwner?.Name);
			}
			else
			{
				GenerateError(writer, $"Unable to use event {member.Member.Name} without a backing class (use x:Class)");
			}
```
So it seems that the `_xClassName` is not set.
In this file it is assigned only once:

```
			var topLevelControl = _fileDefinition.Objects.First();

			BuildNameCache(topLevelControl);

			if (topLevelControl.Type.Name == "ResourceDictionary")
			{
				_isTopLevelDictionary = true;
```
So on, until on the else:
```
			}
			else
			{
				_xClassName = GetClassName(topLevelControl);
```

It seems that the `_xClassName` is not assigned if it is a ResourceDictionary (even if it has a backing class).

Is this intentional or just an accident?
If it is intentional, is there a workaround to use event handlers in ResourceDictionaries - which will build for WinUI, also?

In this project I'm using the latest stable (4.5.12) versions on .NET6, however, the XamlFileGenerator source is current.
It works in UWP, I use this technique there.

**Comment by jeromelaban** [comment]
Thanks for the report. 

This is an unusual use case for code behind. Could you provide a full sample project of what you trying to do, and explain the actual use for this?

**Comment by sarosigy** [comment]
I created a repro, which compiles and runs on Windows, but the skia and wasm head fails to compile:
`Uno.UI.SourceGenerators.XamlGenerator.XamlParsingException: An error was found in Flyout ---> Uno.UI.SourceGenerators.XamlGenerator.XamlParsingException: An error was found in StackPanel ---> Uno.UI.SourceGenerators.XamlGenerator.XamlParsingException: An error was found in CommandBar ---> System.Exception: Unable to find x:Class on the top level element`

It is here: https://github.com/sarosigy/testresdict

The actual use cases (in UWP, I used a lot of similar scenarios):
If a DataTemplate contains for example, a Flyout, which has buttons, the Flyout will not close just by clicking on that buttons. It must be closed by calling Hide. This scenario is implemented in the repro.
A different scenario would be to use controls in a template which has visual states for example, for hover. It must be set in code behind, also, with event handlers.
I use DataTemplates for printing, also. In the "preview" view some elements are clickable (for example, to add a remark to an order's fulfillment, or change the discount percent), and the same template instantiated at printing, from codebehind.

If I use these templates more than one place (for example, display data types with these templates everywhere for consistency) I would put them into a ResourceDictionary which is merged and I can use them by {StaticResource }.

**Reply by jeromelaban** [reply]
Thanks for the update. Let's track this in https://github.com/unoplatform/uno/issues/10109.

**Reply by jeromelaban** [reply]
@sarosigy the behavior shown in the sample is likely the result of an edge case that I'm not able to reproduce when using the Flyout from separate Xaml file, which would avoid relying on what seems to be a side-effect of the XAML parser.

I made [a suggestion](https://github.com/unoplatform/uno/issues/10109#issuecomment-1360318219) on what is think is a better use `Flyout`.
